src/Adapter/Translations/TranslationRouteFinder.php

```diff
@@ -0,0 +1,186 @@
+ * @license   https://opensource.org/licenses/OSL-3.0 Open Software License (OSL 3.0)
+ * International Registered Trademark & Property of PrestaShop SA
+ */
+
+namespace PrestaShop\PrestaShop\Adapter\Translations;
+
+use Link;
+use PrestaShop\PrestaShop\Core\Addon\Module\ModuleRepositoryInterface;
+use PrestaShopBundle\Service\TranslationService;
+use Symfony\Component\HttpFoundation\ParameterBag;
+use Symfony\Component\PropertyAccess\PropertyAccess;
+
+/**
+ * Class TranslationRouteFinder finds the correct route for translations
+ */
+class TranslationRouteFinder
+{
+    /**
+     * @var TranslationService
+     */
+    private $translationService;
+    /**
+     * @var Link
+     */
+    private $link;
+    /**
+     * @var ModuleRepositoryInterface
+     */
+    private $moduleRepository;
+
+    /**
+     * @param TranslationService $translationService
+     * @param Link $link
+     * @param ModuleRepositoryInterface $moduleRepository
+     */
+    public function __construct(
+        TranslationService $translationService,
+        Link $link,
+        ModuleRepositoryInterface $moduleRepository
+    ) {
+        $this->translationService = $translationService;
+        $this->link = $link;
+        $this->moduleRepository = $moduleRepository;
+    }
+
+    /**
+     * Finds the correct translation route out of given query
+     *
+     * @param ParameterBag $query
+     *
+     * @return string
+     */
+    public function findRoute(ParameterBag $query)
+    {
+        $routeProperties = $query->get('modify_translations');
+        $propertyAccessor = PropertyAccess::createPropertyAccessor();
+        $route = 'admin_international_translation_overview';
+
+        switch ($propertyAccessor->getValue($routeProperties, '[translation_type]')) {
+            case 'mails':
+                if ('body' === $propertyAccessor->getValue($routeProperties, '[email_content_type]')) {
+                    $language = $propertyAccessor->getValue($routeProperties, '[language]');
+                    $route = $this->link->getAdminLink(
+                        'AdminTranslations',
+                        true,
+                        [],
+                        [
+                            'lang' => $language,
+                            'type' => 'mails',
+                            'selected-emails' => 'body',
+                            'selected-theme' => $propertyAccessor->getValue($routeProperties, '[theme]'),
+                            'locale' => $this->translationService->langToLocale($language),
+                        ]
+                    );
+                }
+
+                break;
+
+            case 'modules':
+                $moduleName = $propertyAccessor->getValue($routeProperties, '[module]');
+
+                // If module is not using the new translation system -
+                // generate a legacy link for it
+                if (!$this->isModuleUsingNewTranslationSystem($moduleName)) {
```

Comment:
we can see here a good practice: create a meaningful function for a condition :+1: